core/node/api_server/src/tx_sender/gas_estimation.rs

```diff
@@ -441,22 +432,26 @@ impl<'a> GasEstimator<'a> {
     pub(super) async fn unadjusted_step(
         &self,
         forced_gas_limit: u64,
-    ) -> anyhow::Result<(VmExecutionResultAndLogs, TransactionExecutionMetrics)> {
+    ) -> Result<(VmExecutionResultAndLogs, TransactionExecutionMetrics), SubmitTxError> {
         let mut tx = self.transaction.clone();
         match &mut tx.common_data {
             ExecuteTransactionCommon::L1(l1_common_data) => {
                 l1_common_data.gas_limit = forced_gas_limit.into();
-                let required_funds =
-                    l1_common_data.gas_limit * l1_common_data.max_fee_per_gas + tx.execute.value;
+                // Since `tx.execute.value` is supplied by the client and is not checked against the current balance (unlike for L2 transactions),
+                // we may hit an integer overflow. Ditto for protocol upgrade transactions below.
+                let required_funds = (l1_common_data.gas_limit * l1_common_data.max_fee_per_gas)
+                    .checked_add(tx.execute.value)
+                    .ok_or(SubmitTxError::InsufficientFundsForTransfer)?;
```

Comment:
nit: we have several errors with similar meaning:

https://github.com/matter-labs/zksync-era/blob/6dd486555775b3dec0c0de080999a1d73aab25ce/core/node/api_server/src/tx_sender/result.rs#L19-L20C5

https://github.com/matter-labs/zksync-era/blob/6dd486555775b3dec0c0de080999a1d73aab25ce/core/node/api_server/src/tx_sender/result.rs#L35-L36

https://github.com/matter-labs/zksync-era/blob/6dd486555775b3dec0c0de080999a1d73aab25ce/core/node/api_server/src/tx_sender/result.rs#L62-L65

I suppose that we use `InsufficientFundsForTransfer` due to extremely big `tx.execute.value` required to overflow funds calculation. But as stated in the comment, we do not check value against the current balance. This behavior maybe a little bit confusing imo